modules/uploads/utils/bucket.ts

```diff
@@ -0,0 +1,286 @@
+	GetObjectCommand,              // Get URL uploaded file
+	DeleteObjectCommand,           // Delete object
+	DeleteObjectsCommand,          // Delete multiple objects
+	PutObjectCommand,              // Get URL to upload single part file
+	
+	CreateMultipartUploadCommand,   // Initiate a multipart upload
+	UploadPartCommand,            	// Create URL for multipart upload chunk
+	ListPartsCommand,               // List parts of multipart upload
+	CompleteMultipartUploadCommand, // Complete multipart upload
+
+	S3Client,
+} from "npm:@aws-sdk/client-s3";
+import { getSignedUrl } from "npm:@aws-sdk/s3-request-presigner";
+
+import { S3Config } from "./env.ts";
+import { getKey, UploadFile } from "./types.ts";
+import { PresignedUrl } from "./types.ts";
+
+/**
+ * Create a new S3 client instance based on the S3Config
+ * 
+ * @param config S3 Config
+ * @returns A new S3 client
+ */
+export function getClient(config: S3Config) {
+	return new S3Client({
+		region: config.region,
+		credentials: {
+			accessKeyId: config.accessKeyId,
+			secretAccessKey: config.secretAccessKey,
+		},
+		endpoint: config.endpoint,
+		defaultUserAgentProvider: () =>
+			Promise.resolve([
+				["opengb/uploads", "0.1.0"],
+			]),
+	});
+}
+
+/**
+ * Create a presigned URL to perform a SINGLE PART upload to S3
+ * 
+ * @param config S3 Config
+ * @param uploadId The ID of the Upload batch
+ * @param file The file to upload
+ * @param expirySeconds How long the upload URL should be valid for
+ * @returns A URL that exposes an HTTP PUT endpoint to upload the file
+ */
+export async function getPresignedPutUrl(
+	config: S3Config,
+	uploadId: string,
+	file: UploadFile,
+	expirySeconds = 60 * 60,
+): Promise<{ key: string, urls: PresignedUrl[] }> {
+	const client = getClient(config);
+
+	const key = getKey(uploadId, file.path);
+	const command = new PutObjectCommand({
+		Bucket: config.bucket,
+		Key: key,
+		ContentType: file.mime ?? undefined,
+		ContentLength: parseInt(file.contentLength),
+	});
+	const url = await getSignedUrl(
+		client,
+		command,
+		{ expiresIn: expirySeconds },
+	);
+
+	return {
+		urls: [{
+			url,
+			partNumber: 0,
+			offset: "0",
+			contentLength: file.contentLength,
+		}],
+		key,
+	};
+}
+
+export async function getPresignedMultipartUploadUrls(
+	config: S3Config,
+	uploadId: string,
+	file: UploadFile,
+	chunkSize: bigint,
+	expirySeconds = 60 * 60 * 6,
+): Promise<{ key: string, urls: PresignedUrl[], multipartUploadId: string }> {
+	const client = getClient(config);
+
+	const key = getKey(uploadId, file.path);
+	const command = new CreateMultipartUploadCommand({
+		Bucket: config.bucket,
+		Key: key,
+		ContentType: file.mime ?? undefined,
+		
+	});
+	const multipartUpload = await client.send(command);
+
+	const id = multipartUpload.UploadId;
+	if (!id) throw new Error("Multipart upload ID not returned");
+
+	// Round up to the nearest chunk count
+	const chunkCount = (BigInt(file.contentLength) + chunkSize - 1n) / chunkSize;
```

Comment:
Make sure to ensure that there are no more than 10000 parts https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html